esp-hal/src/timer/systimer.rs

```diff
@@ -418,20 +420,22 @@ pub trait Comparator {
     fn set_enable(&self, enable: bool) {
         let systimer = unsafe { &*SYSTIMER::ptr() };
 
-        critical_section::with(|_| {
+        lock(&CONF_LOCK, || {
             #[cfg(not(esp32s2))]
             systimer.conf().modify(|_, w| match self.channel() {
                 0 => w.target0_work_en().bit(enable),
                 1 => w.target1_work_en().bit(enable),
                 2 => w.target2_work_en().bit(enable),
                 _ => unreachable!(),
             });
-
-            #[cfg(esp32s2)]
-            systimer
-                .target_conf(self.channel() as usize)
-                .modify(|_r, w| w.work_en().bit(enable));
         });
+
+        // Note: The ESP32-S2 doesn't require a lock because each
+        // comparator's enable bit in a different register.
+        #[cfg(esp32s2)]
+        systimer
```

Comment:
I'm still not convinced this is sound, as the implementors are still possible to be Sync. If one context modifies the period, and the other context modifies work_en, one of them will have a chance of restoring the other's old state. You should not avoid locking unless the register has a single bit, or the method takes `&mut self`.

Reply:
I'll make the objects not `Sync`, then use from multiple contexts would be illegal.

Reply:
That kinda invalidates this whole PR and steers the users toward "slap it in a Mutex<RefCell>". I would like to not do that, please.

Reply:
The mutex that is used for the non-S2 chips was added because multiple objects, `Unit<0>`, `Unit<1>`, `Comparator<0>`, etc. which could be sent to different threads/contexts, needed to modify the same register.

The mutex wasn't added to allow sharing of the same object. These objects were never intended to be shared, only sent, for the reasons you described "one of them will have a chance of restoring the other's old state".

I definitely agree with you about `Mutex<RefCell>` being bad but I don't see why locking helps in the S2 case.
Even if I were to wrap it in a lock. How many `LockState` objects would be necessary?

Reply:
I'm not stubbornly-refusing-to-accept your arguments here (even if I'm not ready to accept them right now), but in the context of this PR it's not technically a correct change :)